From patchwork Thu Nov  8 09:14:41 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Tianyu Lan <ltykernel@gmail.com>
X-Patchwork-Id: 10673819
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 9E15F109C
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:15:43 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 8D6002CF99
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:15:43 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 7F0372CF9E; Thu,  8 Nov 2018 09:15:43 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No,
 score=-7.7 required=2.0 tests=BAYES_00,DKIM_ADSP_CUSTOM_MED,
	DKIM_INVALID,DKIM_SIGNED,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id F35002CF99
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:15:42 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726856AbeKHSuK (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 8 Nov 2018 13:50:10 -0500
Received: from mail-pl1-f193.google.com ([209.85.214.193]:46252 "EHLO
        mail-pl1-f193.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726133AbeKHSuK (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 8 Nov 2018 13:50:10 -0500
Received: by mail-pl1-f193.google.com with SMTP id c13-v6so9246997plz.13;
        Thu, 08 Nov 2018 01:15:36 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=BZQ//cVaU9tNuNQAUhzd/1Jm0JpGnlnc9pX5EOH8jfE=;
        b=LxIFTCt/GTcyxU0v9nk4twJxEGVigy1BlU2eCQPGqVBOUFZaAlHLG32Ki6yJcGNp/s
         xPuN5EnahHYlwz4TQ5jIcInfJ9az7mLvFgNuxRiU30b6aj2+u/60bs+5Q26rAKVSHhK+
         sNgWw+l55/iP7cf8hrj3jUuSQjqh4ZEdq9r6CKTj9oeYH8xmR56eUgINLkkc1G9SmSic
         ITdNvpD+CaRbBHfsPGhc76nDGi58sL/Tu/DUhBbg6YSGJVEugPvCqoU1jK18V00xzKX0
         +5j8Wj4Uy+5p0MpVftDhSgLtHvILx8dMaugq0NghFzrFbix6ZnUojwtP1vcGr8lb84fj
         JpmQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references;
        bh=BZQ//cVaU9tNuNQAUhzd/1Jm0JpGnlnc9pX5EOH8jfE=;
        b=imph9MOeFCCwNVOAM7qq2kBaT5vDkdFSxWH7whcKEh5/WFpn2djA9WKrNwdJwUigVB
         iRdNc05GPKDyFk6UGxxQtsMmqw0yHxVi7wbbBdEvzXfJD0li6rGaSVWMC/nn9FAk/nZ9
         DiAWcfgddgbqLG4N0OHNGU7OTXkviLN7yOaegVTNn++zd39awA/F3xFKK5BQGCal1fyK
         qm8qkU8P73Y6jnhM93fy8nGR2aSouFTmRvrr5q6KNO146G0aaxajA+CaqgMB5HPSnE0K
         WUzsIy0MgWdRVcmlZjESsczqh7b5kulbNDgsUgkhP5Ox5Y+sCh7cZYeqQDLi77nUiJZk
         DYEg==
X-Gm-Message-State: AGRZ1gJrR1dPL0OrXPXP/mEzKCHkH5PXe86y2ppR7e5cmT3yN19j3Wln
        TIsw4YF5FiQPK4UuZ3Bb7FY=
X-Google-Smtp-Source: 
 AJdET5cjj8OCQlI0d0gch8uRXb7ma45QoRkvqCHNOyOIR5/cXRrSxwKM/LBAsOiq6x1ABZ8iRqn4JQ==
X-Received: by 2002:a17:902:8210:: with SMTP id
 x16-v6mr3812752pln.129.1541668535729;
        Thu, 08 Nov 2018 01:15:35 -0800 (PST)
Received: from localhost.corp.microsoft.com
 ([2404:f801:9000:18:d9bf:62c6:740b:9fc4])
        by smtp.googlemail.com with ESMTPSA id
 w66-v6sm3284114pfb.51.2018.11.08.01.15.28
        (version=TLS1_2 cipher=ECDHE-RSA-CHACHA20-POLY1305 bits=256/256);
        Thu, 08 Nov 2018 01:15:35 -0800 (PST)
From: ltykernel@gmail.com
X-Google-Original-From: Tianyu.Lan@microsoft.com
Cc: Lan Tianyu <Tianyu.Lan@microsoft.com>, christoffer.dall@arm.com,
        marc.zyngier@arm.com, linux@armlinux.org.uk,
        catalin.marinas@arm.com, will.deacon@arm.com, jhogan@kernel.org,
        ralf@linux-mips.org, paul.burton@mips.com, paulus@ozlabs.org,
        benh@kernel.crashing.org, mpe@ellerman.id.au, pbonzini@redhat.com,
        rkrcmar@redhat.com, tglx@linutronix.de, mingo@redhat.com,
        bp@alien8.de, hpa@zytor.com, x86@kernel.org,
        linux-arm-kernel@lists.infradead.org, kvmarm@lists.cs.columbia.edu,
        linux-kernel@vger.kernel.org, linux-mips@linux-mips.org,
        kvm-ppc@vger.kernel.org, linuxppc-dev@lists.ozlabs.org,
        kvm@vger.kernel.org, michael.h.kelley@microsoft.com,
        kys@microsoft.com, vkuznets@redhat.com
Subject: [PATCH V5 4/10] KVM/VMX: Add hv tlb range flush support
Date: Thu,  8 Nov 2018 17:14:41 +0800
Message-Id: <20181108091447.8275-5-Tianyu.Lan@microsoft.com>
X-Mailer: git-send-email 2.14.4
In-Reply-To: <20181108091447.8275-1-Tianyu.Lan@microsoft.com>
References: <20181108091447.8275-1-Tianyu.Lan@microsoft.com>
To: unlisted-recipients:; (no To-header on input)
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Lan Tianyu <Tianyu.Lan@microsoft.com>

This patch is to register tlb_remote_flush_with_range callback with
hv tlb range flush interface.

Signed-off-by: Lan Tianyu <Tianyu.Lan@microsoft.com>
---
Change since v4:
	- Use new function kvm_fill_hv_flush_list_func() to fill flush
       request.
Change since v3:
	- Merge Vitaly's don't pass EPT configuration info to
vmx_hv_remote_flush_tlb() fix.
Change since v1:
	- Pass flush range with new hyper-v tlb flush struct rather
       than KVM tlb flush struct.
---
 arch/x86/kvm/vmx.c | 69 ++++++++++++++++++++++++++++++++++++++----------------
 1 file changed, 49 insertions(+), 20 deletions(-)

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index edbc96cb990a..405dfbde70b2 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -1567,7 +1567,38 @@ static void check_ept_pointer_match(struct kvm *kvm)
 	to_kvm_vmx(kvm)->ept_pointers_match = EPT_POINTERS_MATCH;
 }
 
-static int vmx_hv_remote_flush_tlb(struct kvm *kvm)
+int kvm_fill_hv_flush_list_func(struct hv_guest_mapping_flush_list *flush,
+		void *data)
+{
+	struct kvm_tlb_range *range = data;
+
+	return hyperv_fill_flush_guest_mapping_list(flush, range->start_gfn,
+			range->pages);
+}
+
+static inline int __hv_remote_flush_tlb_with_range(struct kvm *kvm,
+		struct kvm_vcpu *vcpu, struct kvm_tlb_range *range)
+{
+	u64 ept_pointer = to_vmx(vcpu)->ept_pointer;
+
+	/* If ept_pointer is invalid pointer, bypass flush request. */
+	if (!VALID_PAGE(ept_pointer))
+		return 0;
+
+	/*
+	 * FLUSH_GUEST_PHYSICAL_ADDRESS_SPACE hypercall needs address
+	 * of the base of EPT PML4 table, strip off EPT configuration
+	 * information.
+	 */
+	if (range)
+		return hyperv_flush_guest_mapping_range(ept_pointer & PAGE_MASK,
+				kvm_fill_hv_flush_list_func, (void *)range);
+	else
+		return hyperv_flush_guest_mapping(ept_pointer & PAGE_MASK);
+}
+
+static int hv_remote_flush_tlb_with_range(struct kvm *kvm,
+		struct kvm_tlb_range *range)
 {
 	struct kvm_vcpu *vcpu;
 	int ret = -ENOTSUPP, i;
@@ -1577,30 +1608,23 @@ static int vmx_hv_remote_flush_tlb(struct kvm *kvm)
 	if (to_kvm_vmx(kvm)->ept_pointers_match == EPT_POINTERS_CHECK)
 		check_ept_pointer_match(kvm);
 
-	/*
-	 * FLUSH_GUEST_PHYSICAL_ADDRESS_SPACE hypercall needs the address of the
-	 * base of EPT PML4 table, strip off EPT configuration information.
-	 * If ept_pointer is invalid pointer, bypass the flush request.
-	 */
 	if (to_kvm_vmx(kvm)->ept_pointers_match != EPT_POINTERS_MATCH) {
-		kvm_for_each_vcpu(i, vcpu, kvm) {
-			if (!VALID_PAGE(to_vmx(vcpu)->ept_pointer))
-				return 0;
-
-			ret |= hyperv_flush_guest_mapping(
-				to_vmx(vcpu)->ept_pointer & PAGE_MASK);
-		}
+		kvm_for_each_vcpu(i, vcpu, kvm)
+			ret |= __hv_remote_flush_tlb_with_range(
+					kvm, vcpu, range);
 	} else {
-		if (!VALID_PAGE(to_vmx(kvm_get_vcpu(kvm, 0))->ept_pointer))
-			return 0;
-
-		ret = hyperv_flush_guest_mapping(
-			to_vmx(kvm_get_vcpu(kvm, 0))->ept_pointer & PAGE_MASK);
+		ret = __hv_remote_flush_tlb_with_range(kvm,
+				kvm_get_vcpu(kvm, 0), range);
 	}
 
 	spin_unlock(&to_kvm_vmx(kvm)->ept_pointer_lock);
 	return ret;
 }
+
+static int hv_remote_flush_tlb(struct kvm *kvm)
+{
+	return hv_remote_flush_tlb_with_range(kvm, NULL);
+}
 #else /* !IS_ENABLED(CONFIG_HYPERV) */
 static inline void evmcs_write64(unsigned long field, u64 value) {}
 static inline void evmcs_write32(unsigned long field, u32 value) {}
@@ -7957,8 +7981,11 @@ static __init int hardware_setup(void)
 
 #if IS_ENABLED(CONFIG_HYPERV)
 	if (ms_hyperv.nested_features & HV_X64_NESTED_GUEST_MAPPING_FLUSH
-	    && enable_ept)
-		kvm_x86_ops->tlb_remote_flush = vmx_hv_remote_flush_tlb;
+	    && enable_ept) {
+		kvm_x86_ops->tlb_remote_flush = hv_remote_flush_tlb;
+		kvm_x86_ops->tlb_remote_flush_with_range =
+				hv_remote_flush_tlb_with_range;
+	}
 #endif
 
 	if (!cpu_has_vmx_ple()) {
@@ -11567,6 +11594,8 @@ static struct kvm_vcpu *vmx_create_vcpu(struct kvm *kvm, unsigned int id)
 	vmx->nested.posted_intr_nv = -1;
 	vmx->nested.current_vmptr = -1ull;
 
+	vmx->ept_pointer = INVALID_PAGE;
+
 	vmx->msr_ia32_feature_control_valid_bits = FEATURE_CONTROL_LOCKED;
 
 	/*

From patchwork Thu Nov  8 09:14:45 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Tianyu Lan <ltykernel@gmail.com>
X-Patchwork-Id: 10673827
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 8AAB018FD
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:16:22 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 799422CFA7
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:16:22 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 6D4962CFAA; Thu,  8 Nov 2018 09:16:22 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No,
 score=-7.7 required=2.0 tests=BAYES_00,DKIM_ADSP_CUSTOM_MED,
	DKIM_INVALID,DKIM_SIGNED,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 16F862CFA7
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:16:22 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726967AbeKHSuh (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 8 Nov 2018 13:50:37 -0500
Received: from mail-pf1-f196.google.com ([209.85.210.196]:34349 "EHLO
        mail-pf1-f196.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726133AbeKHSuh (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 8 Nov 2018 13:50:37 -0500
Received: by mail-pf1-f196.google.com with SMTP id y18-v6so6646843pfn.1;
        Thu, 08 Nov 2018 01:16:03 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=qnZeUnqhcwg+sgRRpHhl3xeCjMEI8GDTWinkp6r9EUA=;
        b=rYIHXMgSSEM4AVSxH4EPlBjMUhrOqYGTn2alQ9uUS+QxHhe0DoHoJUc4JWMc5zJ963
         yuxLFUZhEJxJdKwRuzQr71FxMFb7Ud7IP3zixx7EVMqplnoJPOEyPIcaGrAY+uxPweBQ
         ay+b8TJ3iVZynk2kf2q0sE6lPpl9IivyLx9zolQCVfOXZym4WDB0z7sDRrmGKKnGr/1C
         CAg8oYIwZ30eg62WNsuiMznmAkgqR/iUJHkbM81uwv+NCGOSBgWPckLg+l+0h3hUT+CY
         ReVxJcUdorTCDaS66st4OQrdUfWcQLzf49HlM1Gmrkb3FscFpaCV9S7noL8rWHARQpUu
         vgMA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references;
        bh=qnZeUnqhcwg+sgRRpHhl3xeCjMEI8GDTWinkp6r9EUA=;
        b=SHvLRtHchG8ctyf6mRfWWPcYayzDX5+fjvMp4eVeeIje9cbv9Xxdi/bKdPaOhiQUR5
         VpZdmJTzu5jjOTFaqKheXJY1eNyhlrnTSStrNW3LrGiRy5q7hmLLEwkjbPez1u0oJnUP
         9RIRTA4ufZ7rHu6ER5PfVC9BZRUy+j+DBv33DM/e9I+b2kJJGtidLVolJCVCXe9VVnL3
         2n8Uk7zlDw5xy1a1KTXp2wDcrWjaidjRNvJ5vsm4wKNgcFhJ9ZMg9ZfGHHwXlE8U6NGB
         4OJunB1AFOSe1up+kldiOuq7Wq/MD2/yVC8jl7Bu0kYn/UJ07Yx7Xidsfk1+t6Aq6lg8
         ukCQ==
X-Gm-Message-State: AGRZ1gJaHOshKPFTONOZm4xCmdlbMI1w9VIyGtyBsSvXj6vSsSXE7oB7
        ggQEK9V9lnfEGBf2oMIKniI=
X-Google-Smtp-Source: 
 AJdET5e/YfsxrqewWUNch5xifPripCFpe9eaVa/+bH3/JZoQDnUMFLB9dz3x7nsjPS6Ux8JquGpGAA==
X-Received: by 2002:a62:3c6:: with SMTP id
 189-v6mr3783054pfd.19.1541668562678;
        Thu, 08 Nov 2018 01:16:02 -0800 (PST)
Received: from localhost.corp.microsoft.com
 ([2404:f801:9000:18:d9bf:62c6:740b:9fc4])
        by smtp.googlemail.com with ESMTPSA id
 w66-v6sm3284114pfb.51.2018.11.08.01.15.58
        (version=TLS1_2 cipher=ECDHE-RSA-CHACHA20-POLY1305 bits=256/256);
        Thu, 08 Nov 2018 01:16:02 -0800 (PST)
From: ltykernel@gmail.com
X-Google-Original-From: Tianyu.Lan@microsoft.com
Cc: Lan Tianyu <Tianyu.Lan@microsoft.com>, pbonzini@redhat.com,
        rkrcmar@redhat.com, tglx@linutronix.de, mingo@redhat.com,
        bp@alien8.de, hpa@zytor.com, x86@kernel.org, kvm@vger.kernel.org,
        linux-kernel@vger.kernel.org, michael.h.kelley@microsoft.com,
        kys@microsoft.com, vkuznets@redhat.com, linux@armlinux.org.uk
Subject: [PATCH V5 8/10] KVM/MMU: Move tlb flush in kvm_set_pte_rmapp() to
 kvm_mmu_notifier_change_pte()
Date: Thu,  8 Nov 2018 17:14:45 +0800
Message-Id: <20181108091447.8275-9-Tianyu.Lan@microsoft.com>
X-Mailer: git-send-email 2.14.4
In-Reply-To: <20181108091447.8275-1-Tianyu.Lan@microsoft.com>
References: <20181108091447.8275-1-Tianyu.Lan@microsoft.com>
To: unlisted-recipients:; (no To-header on input)
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Lan Tianyu <Tianyu.Lan@microsoft.com>

This patch is to move tlb flush in kvm_set_pte_rmapp() to
kvm_mmu_notifier_change_pte() in order to avoid redundant tlb flush.

Signed-off-by: Lan Tianyu <Tianyu.Lan@microsoft.com>
---
 arch/x86/kvm/mmu.c  | 8 ++------
 virt/kvm/kvm_main.c | 5 ++++-
 2 files changed, 6 insertions(+), 7 deletions(-)

diff --git a/arch/x86/kvm/mmu.c b/arch/x86/kvm/mmu.c
index 06bfcd327ef6..b13b419166c4 100644
--- a/arch/x86/kvm/mmu.c
+++ b/arch/x86/kvm/mmu.c
@@ -1776,10 +1776,7 @@ static int kvm_set_pte_rmapp(struct kvm *kvm, struct kvm_rmap_head *rmap_head,
 		}
 	}
 
-	if (need_flush)
-		kvm_flush_remote_tlbs(kvm);
-
-	return 0;
+	return need_flush;
 }
 
 struct slot_rmap_walk_iterator {
@@ -1915,8 +1912,7 @@ int kvm_unmap_hva_range(struct kvm *kvm, unsigned long start, unsigned long end)
 
 int kvm_set_spte_hva(struct kvm *kvm, unsigned long hva, pte_t pte)
 {
-	kvm_handle_hva(kvm, hva, (unsigned long)&pte, kvm_set_pte_rmapp);
-	return 0;
+	return kvm_handle_hva(kvm, hva, (unsigned long)&pte, kvm_set_pte_rmapp);
 }
 
 static int kvm_age_rmapp(struct kvm *kvm, struct kvm_rmap_head *rmap_head,
diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 2679e476b6c3..dac4b0446aed 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -354,7 +354,10 @@ static void kvm_mmu_notifier_change_pte(struct mmu_notifier *mn,
 	idx = srcu_read_lock(&kvm->srcu);
 	spin_lock(&kvm->mmu_lock);
 	kvm->mmu_notifier_seq++;
-	kvm_set_spte_hva(kvm, address, pte);
+
+	if (kvm_set_spte_hva(kvm, address, pte))
+		kvm_flush_remote_tlbs(kvm);
+
 	spin_unlock(&kvm->mmu_lock);
 	srcu_read_unlock(&kvm->srcu, idx);
 }

From patchwork Thu Nov  8 09:14:46 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Tianyu Lan <ltykernel@gmail.com>
X-Patchwork-Id: 10673825
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 32788109C
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:16:22 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 1DF032CFA9
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:16:22 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 0C46F2CFAA; Thu,  8 Nov 2018 09:16:22 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No,
 score=-7.7 required=2.0 tests=BAYES_00,DKIM_ADSP_CUSTOM_MED,
	DKIM_INVALID,DKIM_SIGNED,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id A30972CFA9
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:16:21 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727005AbeKHSul (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 8 Nov 2018 13:50:41 -0500
Received: from mail-pf1-f195.google.com ([209.85.210.195]:42462 "EHLO
        mail-pf1-f195.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726133AbeKHSul (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 8 Nov 2018 13:50:41 -0500
Received: by mail-pf1-f195.google.com with SMTP id u10-v6so3289798pfn.9;
        Thu, 08 Nov 2018 01:16:07 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=Mf4jQKDBupERk9IP8j/c7DldYt9kVutHpnKEjeY1Ldg=;
        b=qFvlUnTwf8mkn8q8geC7QphecA07wjvdx/2EoV7iCtwNLDma89hkjq3gNjwUcl8FLW
         vXW1+7m8un2xZvK8GRmj/aNpMvykLO3BqdSaN3UbRsqcdRRE6u0y2Z2vXMpk+McsVqDy
         wM8iP61ovMgwB+JwX3AX9DBwJhJ1Pfpi2NkODQwIKG0E756yloHLGHzmBj1+VbcEttsJ
         z4fKwivTpDKirZKOCEzB11mAuiAd+wFOVojW4gsJaEYRRJl6GXQI7CDcP1zVU3FboSpH
         jZI3cDcbHhPvAbDmwbxfxVINsRQHJie8Nfh0iHogG7PKK85M0hpnj6deyIUQeCIvjpg5
         +9rA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references;
        bh=Mf4jQKDBupERk9IP8j/c7DldYt9kVutHpnKEjeY1Ldg=;
        b=lD7oPQxokhflxtOMhjkMI/Vn1eBgWR3JfXdxZCQKldU72k8aUmLqIdlYWo69biaQb8
         awQakyQbY89DKADE57weTXf1P+LHuGs/7oVi+5/kipXccbTpSBvnEhsPzCnXy8xk2Zdj
         KBJTe3/sWkAsnuJx9pZX5NdqlkYdhkV4DORoAuaOFUOSdjb/RyYafK8lIeceEkhzJmu4
         GTy16qWTGZuKyNjfRWcRiLjvWPDepmqyJVmhoOiKCu8z5ZuZ36dNP+FV7CDuIQQsDDGr
         jYI4m9Kuz1rP2V0GR+qIhVGL3LOMZerbdU8ylEd41MitGvLupIhQQ+vwB4nRv8hr9hML
         vjjQ==
X-Gm-Message-State: AGRZ1gKhC1r99mOziY+iz57QCsyf86TJjoe+Ge4og6nG9k+Mi5Whb/Ke
        CBQe4I/zWZk4Tgs5kj+d5ec=
X-Google-Smtp-Source: 
 AJdET5erW+aY8VP9uRwInYHOrvptDypjA6boOWPxTG4eRUnWTKpQG0OPxeYjO5VPecyQrdHnQ0YbWA==
X-Received: by 2002:a62:da54:: with SMTP id
 w20-v6mr3793532pfl.106.1541668567203;
        Thu, 08 Nov 2018 01:16:07 -0800 (PST)
Received: from localhost.corp.microsoft.com
 ([2404:f801:9000:18:d9bf:62c6:740b:9fc4])
        by smtp.googlemail.com with ESMTPSA id
 w66-v6sm3284114pfb.51.2018.11.08.01.16.02
        (version=TLS1_2 cipher=ECDHE-RSA-CHACHA20-POLY1305 bits=256/256);
        Thu, 08 Nov 2018 01:16:06 -0800 (PST)
From: ltykernel@gmail.com
X-Google-Original-From: Tianyu.Lan@microsoft.com
Cc: Lan Tianyu <Tianyu.Lan@microsoft.com>, pbonzini@redhat.com,
        rkrcmar@redhat.com, tglx@linutronix.de, mingo@redhat.com,
        bp@alien8.de, hpa@zytor.com, x86@kernel.org, kvm@vger.kernel.org,
        linux-kernel@vger.kernel.org, michael.h.kelley@microsoft.com,
        kys@microsoft.com, vkuznets@redhat.com, linux@armlinux.org.uk
Subject: [PATCH V5 9/10] KVM/MMU: Flush tlb directly in the
 kvm_set_pte_rmapp()
Date: Thu,  8 Nov 2018 17:14:46 +0800
Message-Id: <20181108091447.8275-10-Tianyu.Lan@microsoft.com>
X-Mailer: git-send-email 2.14.4
In-Reply-To: <20181108091447.8275-1-Tianyu.Lan@microsoft.com>
References: <20181108091447.8275-1-Tianyu.Lan@microsoft.com>
To: unlisted-recipients:; (no To-header on input)
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Lan Tianyu <Tianyu.Lan@microsoft.com>

This patch is to flush tlb directly in the kvm_set_pte_rmapp()
and return 0.

Signed-off-by: Lan Tianyu <Tianyu.Lan@microsoft.com>
---
 arch/x86/kvm/mmu.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/arch/x86/kvm/mmu.c b/arch/x86/kvm/mmu.c
index b13b419166c4..39e0e2572710 100644
--- a/arch/x86/kvm/mmu.c
+++ b/arch/x86/kvm/mmu.c
@@ -1776,6 +1776,11 @@ static int kvm_set_pte_rmapp(struct kvm *kvm, struct kvm_rmap_head *rmap_head,
 		}
 	}
 
+	if (need_flush && kvm_available_flush_tlb_with_range()) {
+		kvm_flush_remote_tlbs_with_address(kvm, gfn, 1);
+		return 0;
+	}
+
 	return need_flush;
 }
 

From patchwork Thu Nov  8 09:14:47 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Tianyu Lan <ltykernel@gmail.com>
X-Patchwork-Id: 10673823
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 3CC0715E9
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:16:14 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 2CAC52CFA7
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:16:14 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 2101A2CFAA; Thu,  8 Nov 2018 09:16:14 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No,
 score=-7.7 required=2.0 tests=BAYES_00,DKIM_ADSP_CUSTOM_MED,
	DKIM_INVALID,DKIM_SIGNED,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 0D91B2CFA7
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  8 Nov 2018 09:16:13 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726938AbeKHSup (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 8 Nov 2018 13:50:45 -0500
Received: from mail-pl1-f196.google.com ([209.85.214.196]:34657 "EHLO
        mail-pl1-f196.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726133AbeKHSup (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 8 Nov 2018 13:50:45 -0500
Received: by mail-pl1-f196.google.com with SMTP id f12-v6so8078252plo.1;
        Thu, 08 Nov 2018 01:16:11 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=XvvPy7Q4qFhIcL2nJjG34LofSZ7R5VeljzNsh5L5Nb8=;
        b=DEhtQBqXnAEIMkbcf2z0yIj3vhQ93LgYqzJ9qlXT6XP4/fd490GehyvO8sTpfp7xJ8
         KrNclJnEWr09b7CucahuuqcGeVxXbdc1+vF5wbKD2AFgokvyV5srkExxl1lRAOgFrt4a
         jTVkTNP3Oo/N3p53aswcO2uEAEdqRJtNZMOTBLaC7ntAdCdY6p4wOadjoNcZpW/k36dc
         ypo6F4fNyzgkEP7XNz7ZIgxM8ESluYpM0DCRS+G5kySr+FdFl6XLVl+FCqr3CedhSgw2
         naOiSDsCu5yv2LBNSbb3qA62ZvmckieMUygEfuIg56E817/yPIlTNiJYELpsDDySc/fF
         kBwQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references;
        bh=XvvPy7Q4qFhIcL2nJjG34LofSZ7R5VeljzNsh5L5Nb8=;
        b=RThjUuK95zgRWQuF4ALUg39A17R0r5stcmnka4S9c8RQiuti2wOXRH1JHzk0OYuOG0
         3EyKgjTdKEdqqIPVwlDjPpqf4c/GCCJRBbURGCXFLbwQo+RNozNtkPIb2kbRVCAJgaAd
         PL5jRCkR8FktBf99LvbW3gmUx/4PqijikfGZoFAFosQ6WEwpesAl3Smt4mPrW2FuYXzn
         apxDfVDHjSr9+nWZ0feb242LoI+MTbAXmVN+KVcMeUSti9KCsCc9wkGl6R4Cu/F8r019
         nawff4KRspPbDGIirAUgR2osq3xqrVwWos1u4Zm7C19C07m1RxmO69wdz1na+Um9r5eQ
         H8/g==
X-Gm-Message-State: AGRZ1gJTNwpJziHAegOGYbkQizqDgSkcwWXSqaYTEDGH80DXIaryT1MD
        bJhbIt7xDq/tH0/7coTherE=
X-Google-Smtp-Source: 
 AJdET5cTpSSOy5lYBlq8QiWi/JCLzDziKFJCGQlK6RGhGZBDjaTSK2Qk1X1Wh+zJ10tF2u4eRMN/Ig==
X-Received: by 2002:a17:902:1122:: with SMTP id
 d31-v6mr3801300pla.259.1541668571406;
        Thu, 08 Nov 2018 01:16:11 -0800 (PST)
Received: from localhost.corp.microsoft.com
 ([2404:f801:9000:18:d9bf:62c6:740b:9fc4])
        by smtp.googlemail.com with ESMTPSA id
 w66-v6sm3284114pfb.51.2018.11.08.01.16.07
        (version=TLS1_2 cipher=ECDHE-RSA-CHACHA20-POLY1305 bits=256/256);
        Thu, 08 Nov 2018 01:16:11 -0800 (PST)
From: ltykernel@gmail.com
X-Google-Original-From: Tianyu.Lan@microsoft.com
Cc: Lan Tianyu <Tianyu.Lan@microsoft.com>, pbonzini@redhat.com,
        rkrcmar@redhat.com, tglx@linutronix.de, mingo@redhat.com,
        bp@alien8.de, hpa@zytor.com, x86@kernel.org, kvm@vger.kernel.org,
        linux-kernel@vger.kernel.org, michael.h.kelley@microsoft.com,
        kys@microsoft.com, vkuznets@redhat.com, linux@armlinux.org.uk
Subject: [PATCH V5 10/10] KVM/MMU: Flush tlb directly in the
 kvm_zap_gfn_range()
Date: Thu,  8 Nov 2018 17:14:47 +0800
Message-Id: <20181108091447.8275-11-Tianyu.Lan@microsoft.com>
X-Mailer: git-send-email 2.14.4
In-Reply-To: <20181108091447.8275-1-Tianyu.Lan@microsoft.com>
References: <20181108091447.8275-1-Tianyu.Lan@microsoft.com>
To: unlisted-recipients:; (no To-header on input)
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Lan Tianyu <Tianyu.Lan@microsoft.com>

Originally, flush tlb is done by slot_handle_level_range(). This patch
is to flush tlb directly in the kvm_zap_gfn_range() when range
flush is available.

Signed-off-by: Lan Tianyu <Tianyu.Lan@microsoft.com>
---
Change since v4:
	Move operation of setting flush_tlb out of for loop.
---
 arch/x86/kvm/mmu.c | 16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git a/arch/x86/kvm/mmu.c b/arch/x86/kvm/mmu.c
index 39e0e2572710..898560b0807c 100644
--- a/arch/x86/kvm/mmu.c
+++ b/arch/x86/kvm/mmu.c
@@ -5633,8 +5633,13 @@ void kvm_zap_gfn_range(struct kvm *kvm, gfn_t gfn_start, gfn_t gfn_end)
 {
 	struct kvm_memslots *slots;
 	struct kvm_memory_slot *memslot;
+	bool flush_tlb = true;
+	bool flush = false;
 	int i;
 
+	if (kvm_available_flush_tlb_with_range())
+		flush_tlb = false;
+
 	spin_lock(&kvm->mmu_lock);
 	for (i = 0; i < KVM_ADDRESS_SPACE_NUM; i++) {
 		slots = __kvm_memslots(kvm, i);
@@ -5646,12 +5651,17 @@ void kvm_zap_gfn_range(struct kvm *kvm, gfn_t gfn_start, gfn_t gfn_end)
 			if (start >= end)
 				continue;
 
-			slot_handle_level_range(kvm, memslot, kvm_zap_rmapp,
-						PT_PAGE_TABLE_LEVEL, PT_MAX_HUGEPAGE_LEVEL,
-						start, end - 1, true);
+			flush |= slot_handle_level_range(kvm, memslot,
+					kvm_zap_rmapp, PT_PAGE_TABLE_LEVEL,
+					PT_MAX_HUGEPAGE_LEVEL, start,
+					end - 1, flush_tlb);
 		}
 	}
 
+	if (flush)
+		kvm_flush_remote_tlbs_with_address(kvm, gfn_start,
+				gfn_end - gfn_start + 1);
+
 	spin_unlock(&kvm->mmu_lock);
 }
 
